File: src/main/java/com/raychenon/leetcode/PartitionLabels.java
package com.raychenon.leetcode;

import java.util.HashMap;
import java.util.LinkedList;
import java.util.List;
import java.util.Map;

/**
 * User: raychenon
 * Date: 4/9/2020
 * https://leetcode.com/problems/partition-labels/
 */
public class PartitionLabels {

    /**
     * Time Complexity O(n)
     * Space Complexity O(1)
     * @param S
     * @return
     */
    public List<Integer> partitionLabels(String S) {
        Map<Character, Integer> last = new HashMap<Character, Integer>();
        for (int i = 0; i < S.length(); i++) {
            last.put(S.charAt(i), i);
        }

        int j = 0;
        int anchor = 0;
        List<Integer> list = new LinkedList<>();
        for (int i = 0; i < S.length(); i++) {
            j = Math.max(j, last.getOrDefault(S.charAt(i), 0));
            if (i == j) {
                list.add(i - anchor + 1);
                anchor = i + 1;
            }
        }
        return list;
    }
}

File: src/main/kotlin/com/raychenon/kotlin/leetcode/PartitionLabels.kt
package com.raychenon.kotlin.leetcode

/**
 * User: raychenon
 * Date: 5/9/2020
 * https://leetcode.com/problems/partition-labels
 */
object PartitionLabels {

    /**
     * Greedy algorithm
     * Time Complexity O(n)
     * Space Complexity O(1)
     */
    fun partitionLabels(S: String): List<Int> {
        val last = IntArray(26)
        for (i in 0 until S.length) {
            last[S[i] - 'a'] = i
        }

        var j = 0
        var start = 0
        val list = mutableListOf<Int>()
        for (i in 0 until S.length) {
            j = maxOf(j, last[S[i] - 'a'])
            if (i == j) {
                val distance = i - start + 1
                list.add(distance)
                start = i + 1
            }
        }
        return list
    }

    /**
     * Time Complexity O(n)
     * Space Complexity O(1)
     */
    fun partitionLabelsOriginalSolution(S: String): List<Int> {
        val n = S.length
        val letMap = mutableMapOf<Char, Letter>()
        for (i in 0 until n) {
            val c = S[i]
            if (letMap.containsKey(c)) {
                letMap[c]!!.end = i
            } else {
                letMap.put(c, Letter(i, i))
            }
        }

//         for((l,v) in letMap){
//             println("v = ${v}")
//         }

        val sizeList = mutableListOf<Int>()

        // map sorted
        var start = 0
        var end = 0
        for (letter in letMap.values) {
            start = minOf(start, letter.start)

            if (letter.start > end) {
                val distance = end - start + 1
                //println("start > end , $end - $start = $distance")
                sizeList.add(distance)
                start = letter.start
            }
            end = maxOf(end, letter.end)
        }

        val distance = end - start + 1
        sizeList.add(distance)

        return sizeList
    }
}


data class Letter(
    val start: Int,
    var end: Int
)